### PR TITLE
Clarify use creating and verifying TPM attestation statements.  

### LINKED ISSUE #1925: TPM attestation verification steps inconsistent with FIDO conformance testing tool
## Proposed Change

Note: This may end up resulting in a WebAuthn spec change (I think it should - read on), or in a close with feedback to the FIDO alliance about the validity of this test in the conformance testing tool, but the intent of opening this issue is to seek a point of view and clarification from experts in the field of TPM attestation for FIDO, then get the spec aligned with our decision.

There is a very confusing part of the TPM specifications related to which hash algorithm to use when computing the digest of the _pubArea_ for verification of the attested name of the _certInfo_ structure. Adding to this confusion is a completely different interpretation between the way WebAuthn is currently written, and the implementation of a TPM attestation test in the FIDO conformance test tool. 

Currently the WebAuthn TPM attestation verification steps [in section 8.3](https://w3c.github.io/webauthn/#sctn-tpm-attestation) include:

`Verify that attested contains a TPMS_CERTIFY_INFO structure as specified in [[TPMv2-Part2]](https://w3c.github.io/webauthn/#biblio-tpmv2-part2) section 10.12.3, whose name field contains a valid Name for pubArea, as computed using the algorithm in the nameAlg field of pubArea using the procedure specified in [[TPMv2-Part1]](https://w3c.github.io/webauthn/#biblio-tpmv2-part1) section 16.`

I draw attention to the words: **using the algorithm in the nameAlg field of pubArea**

Some sources, including [these instructions](https://medium.com/webauthnworks/verifying-fido-tpm2-0-attestation-fc7243847498) documented by @herrjemand indicate that the hash algorithm does *not* come from the _nameAlg_ field of _pubArea_, but instead comes from the TPM_ALG_ID portion of the attested name in the TPMS_CERTIFY_INFO structure. To me this makes complete sense - why wouldn't you co-locate the hash algorithm that was used alongside the hash digest?! Its not logical at all why the hash algorithm to use in the attested name within the TPMS_CERTIFY_INFO structure would come from the _nameAlg_ in _pubArea_ (as currently written in WebAuthn).

In practice (i.e. in Windows TPM attestations observed in the field) these two hash algorithm identifiers are typically set to the same hash algorithm, which means regardless of which interpretation you use, verification of real-world TPM attestations works.

The FIDO conformance test tool however contains a specific test case where these algorithm identifiers are set to different values, to ensure that RPs use the algorithm identifier from the attested name in _certInfo_. If an RP strictly follows the process as currently defined in the WebAuthn specification, then that test case will fail. 

Adding to the confusion is the procedure specified in [[TPMv2-Part1]](https://w3c.github.io/webauthn/#biblio-tpmv2-part1) section 16, from which I draw two observations:
image

1. This seems fairly clear - use the hash algorithm in the name structure itself, which is consistent with the instructions from the aforementioned blog article and the FIDO conformance tool test.
2. This intimates that the nameAlg comes from the nameAlg parameter *in* the publicArea object. I actually think if the word `in` was changed to the word `of` here in the TPM spec, then things wouldn't be so ambiguous.
 
  
So which is correct? 

If using the _nameAlg_ from _pubArea_ is correct, then both the verification information from @herrjemand in his [blog article](https://medium.com/webauthnworks/verifying-fido-tpm2-0-attestation-fc7243847498) is wrong, as is the validity of that specific FIDO conformance test, as is the first observation from the screenshot of section 16 of the TPM spec. Further, its unclear what an RP should do if the _TPM_ALG_ID_ in the attested name is different from that in _pubArea_ because it would make the former redundant.

If using the algorithm identifier from the _TPM_ALG_ID_ in the attested name in _certInfo_ is correct, then the WebAuthn specification should be re-worded, for example:

Verify that attested contains a TPMS_CERTIFY_INFO structure as specified in [[TPMv2-Part2]](https://w3c.github.io/webauthn/#biblio-tpmv2-part2) section 10.12.3, whose name field contains a valid Name for pubArea, as computed using the procedure specified in [[TPMv2-Part1]](https://w3c.github.io/webauthn/#biblio-tpmv2-part1) section 16. Note that the hash algorithm is included within the attested name field of the TPMS_CERTIFY_INFO structure.

Personally I believe it should be the latter, and am happy to pull together a PR with this change in it. It makes no sense to *not* use the hash algorithm identifier that is co-located in the attested name within the TPMS_CERTIFY_INFO structure.

### PR DESCRIPTION
**Pull request by mwiseman-byid**
Closes #1925 

Addressed potential implementation errors in TPM Attestation

Implementation commitment:

Addressed potential implementation errors in TPM Attestation
Addressed potential privacy issues

Documentation and checks

- [x] Affects privacy
- [ ] Affects security
- [ ] Updated explainer ([link](https://github.com/w3c/webauthn/wiki)

***
<a href="https://pr-preview.s3.amazonaws.com/mwiseman-byid/webauthn/pull/2193.html" title="Last updated on Nov 13, 2024, 3:45 PM UTC (9618b97)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/webauthn/2193/406ec42...mwiseman-byid:9618b97.html" title="Last updated on Nov 13, 2024, 3:45 PM UTC (9618b97)">Diff</a>

### CODE REVIEW

File: index.bs [hunk 1]
```diff
@@ -6536,8 +6536,11 @@ engine.
     setting the `extraData` parameter to the digest of |attToBeSigned| using the hash algorithm corresponding to the "alg" signature algorithm.
     (For the "RS256" algorithm, this would be a SHA-256 digest.)
 
-    Set the |pubArea| field to the public area of the credential public key, the |certInfo| field to the output parameter of the
-    same name, and the |sig| field to the signature obtained from the above procedure.
+    Set the |pubArea| field to the public area of the credential public key (the TPMT_PUBLIC structure), the |certInfo| field (the TPMS_ATTEST structure)
+    to the output parameter of the same name, and the |sig| field to the signature obtained from the above procedure.
+
+    Note: If the |pubArea| is read from the TPM using the TPM2_ReadPublic command, that command returns a TPM2B_PUBLIC structure. TPM2B_PUBLIC
```
Comment (sbweeden) [inline index.bs:6542]: This Note does not seem relevant to RP processing, which what this section is really all about.

**Review by sbweeden — CHANGES_REQUESTED** [review]
Few suggestions inline, open for discussion.

File: index.bs (hunk 1 above)
Comment (emlun) [reply index.bs:6542]: Not really - this is in the signing procedure, which is performed by the authenticator. The RP part is below in the verification procedure.

**Comment by simoneonofri** [comment]
@mwiseman-byid, you should have received an email from IPR Bot asking for your commitment to this PR, as you're not part of this specific W3C Group.

I am available if you need more information.

File: index.bs (hunk 1 above)
Comment (sbweeden) [reply index.bs:6542]: Fair - withdrawn.

**Comment by simoneonofri** [comment]
@mwiseman-byid, the bot is happy. So, from IPR perspective, we're done.

**Comment by nadalin** [comment]
thank you

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

File: index.bs [hunk 2]
```diff
@@ -6551,24 +6554,30 @@ engine.
...
     - Verify that `attested` contains a `TPMS_CERTIFY_INFO` structure as specified in [[!TPMv2-Part2]] section 10.12.3,
-        whose `name` field contains a valid Name for |pubArea|,
-        as computed using the procedure specified in [[!TPMv2-Part1]]
-        section 16. Note that the hash algorithm is included within the attested `name` field of the TPMS_CERTIFY_INFO structure.
-    - Verify that |x5c| is present.
-    - Note that the remaining fields in the "Standard Attestation Structure" [[!TPMv2-Part1]]
+        whose `name` field contains a valid Name for |pubArea|, as computed using the procedure specified in [[!TPMv2-Part1]] section 16 using the nameAlg in the |pubArea|.
+
+        Note: The hash algorithm is also included within the attested `name`
+        field of the TPMS_CERTIFY_INFO structure and will also match nameAlg in |pubArea| when returned by the TPM.
```
[the remark this answers is not in the document]
Comment (mwiseman-byid) [reply index.bs:6594]: By making this a note, this is an informative statement so not sure if we can make this normative. This clarification, in fact, is what I believe is the source of the original concern that there was a possibility these would be different, when, in fact, the TPM never changes the algorithm. Making this check a MUST alludes that the TPM might return these different. That said, as the algId in the Name is not protected it is possible for software (local or in transit) to change it but the worse case would be a DOS attack (unless someone finds a pre-image attack). With that said, I'm open to the suggestion if you still believe that's needed.

[the remark this answers is not in the document]
Comment (sbweeden) [reply index.bs:6594]: The problem I have with lowercase "will also match" is that it's suggestive, without providing guidance to the RP as to what to do if its observed that the values are not the same in both locations. I'm not going to die on this hill though.

File: index.bs [hunk 3]
```diff
@@ -6551,24 +6554,30 @@ engine.
...
+
+        Note: The remaining fields in the "Standard Attestation Structure" [[!TPMv2-Part1]]
         section 31.2, i.e., `qualifiedSigner`, `clockInfo` and `firmwareVersion` are ignored.
-        These fields MAY be used as an input to risk engines.
+        Depending on the properties of the |aikCert| key used, these fields may be obfuscated.
+        If valid, these MAY be used as an input to risk engines.
```
[the remark this answers is not in the document]
Comment (mwiseman-byid) [reply index.bs:6599]: The content of these fields is determined by which TPM hierarchy the AK is in.   (tl;dr: If the AK is in the Endorsement hierarchy the fields are valid, if the AK is in the Storage hierarchy, the fields are random - this is for privacy reasons). To provide information about where the AK is may be a privacy concern so this is not provided in the protocol. The properties of the AK (including whether to even trust it) are out of scope for this protocol and it would be up to the RP to know that out of band.

File: index.bs [hunk 4]
```diff
@@ -6582,6 +6591,11 @@ TPM [=attestation certificate=] MUST have the following fields/extensions:
 
 - The Subject Alternative Name extension MUST be set as defined in [[!TPMv2-EK-Profile]] section 3.2.9.
 
+        Note: Previous versions of [[!TPMv2-EK-Profile]] allowed the inclusion of an optional attribute,
+        called HardwareModuleName, that contains the TPM serial number in the EK certificate.
+        HardwareModuleName SHOULD NOT be placed in in the [=attestation certificate=]
```
[the remark this answers is not in the document]
Comment (mwiseman-byid) [reply index.bs:6616]: Agree MUST NOT would be my preference. However, my concern with MUST NOT is if there are any existing implementation that followed the original wording, MUST NOT would make those existing implementations (if >0) non-compliant which would not be their fault. I'm OK with the guidance of the WG on this.

**Review by sbweeden — APPROVED** [review]

**Review by emlun — APPROVED** [review]
